Utils/MessageUtils.cs

```diff
@@ -1,106 +1,119 @@
             ReadWriteUtils.Arr_Insert(msg, 0, msg.Length, MMFileList[fileIndex].Data, addr);
         }
 
-        private static MMMesssage FindMesssage(int address)
+        private static Dictionary<ushort, MessageEntry> GetMessageTable()
         {
-            int fileIndex = RomUtils.GetFileIndexForWriting(Addresses.TextTable);
-            int baseAddress = Addresses.TextTable - MMFileList[fileIndex].Addr;
-
-            MMMesssage message = new MMMesssage();
+            Dictionary<ushort, MessageEntry> messageTable = new Dictionary<ushort, MessageEntry>();
 
+            int fileIndex = RomUtils.GetFileIndexForWriting(MESSAGE_TABLE_ADDRESS);
+            MMFile file = MMFileList[fileIndex];
+            int baseAddress = MESSAGE_TABLE_ADDRESS - file.Addr;
+            var data = file.Data;
+
             while (true)
             {
-                int x = (MMFileList[fileIndex].Data[baseAddress] << 8)
-                    + MMFileList[fileIndex].Data[baseAddress + 1];
-
-                if (address == x)
-                {
-                    var data = MMFileList[fileIndex].Data;
-                    message.Address = (int)(ReadWriteUtils.Arr_ReadU32(data, baseAddress + 4) & 0xFFFFFF);
-                    message.Size = (int)(ReadWriteUtils.Arr_ReadU32(data, baseAddress + 12) & 0xFFFFFF) - message.Address;
-                    break;
-                }
-
-                if (x > address)
-                {
-                    return null;
-                }
-
+                ushort textId = ReadWriteUtils.Arr_ReadU16(data, baseAddress);
+                if (textId >= 0xFFFD) //This id is still valid, but hard to determine it's size
+                {
+                    break;
+                }
+
+                int address = ReadWriteUtils.Arr_ReadS32(data, baseAddress + 4) & 0xFFFFFF;
+                int addressNext = ReadWriteUtils.Arr_ReadS32(data, baseAddress + 12) & 0xFFFFFF;
+
+                MessageEntry message = new MessageEntry()
+                {
+                    Id = textId,
+                    Address = address,
+                    Size = addressNext - address
+                };
+                messageTable.Add(textId, message);
                 baseAddress += 8;
             }
 
-            return message;
+            return messageTable;
         }
 
-        public static bool IsBadMesssage(string message)
+        private static bool IsBadMessage(string message)
         {
             return message.Contains("a segment of health") || message.Contains("currency") ||
                 message.Contains("money") || message.Contains("cash") ||
                 message.Contains("wealth") || message.Contains("riches and stuff") ||
                 message.Contains("increased life");
         }
 
-        public static void WriteGossipMessage(List<string> messages, Random RNG)
+        public static void WriteGossipMessage(List<string> hints, Random random)
```

Comment:
If messages are always hints, maybe rename the function:

`WriteGossipHints(List<string> hints, Random random)`